44414 data frame attributes not propagating with astype

skwirskj opened the pull request:
- [x] closes #44414 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

skwirskj committed:
BUG: GH#44414 Added __finalize__ call

Added a finalize call so that the metadata, specifically attributes, is retained when the datatype of one column is changed.

skwirskj committed:
BUG: GH#44414 Added comments to indicate which issue was fixed

skwirskj committed:
BUG: GH$44414 Fixed finalize parameter and added test

Passed previous DataFrame to finalize to ensure the correct attributes are being passed to the new DataFrame. Also added a test to test that calling astype passes attributes to the new DataFrame.

skwirskj committed:
Merge remote-tracking branch 'upstream/master' into 44414-DataFrameAttr

skwirskj committed:
BUG: GH#44414 Updated whatsnew with bug fix

phofl commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -636,7 +636,7 @@ Conversion
 - Bug in :func:`to_datetime` with ``arg:xr.DataArray`` and ``unit="ns"`` specified raises TypeError (:issue:`44053`)
 - Bug in :meth:`DataFrame.convert_dtypes` not returning the correct type when a subclass does not overload :meth:`_constructor_sliced` (:issue:`43201`)
--
+- Bug in :meth:`NDFrame.astype` not propagating ``attrs`` from the original :class:`DataFrame` (:issue:`44414`)
```
Please use DataFrame not NDFrame

skwirskj replied:
Got it!

phofl commented on pandas/core/generic.py:
```diff
@@ -5890,6 +5890,8 @@ def astype(
         # GH 19920: retain column metadata after concat
         result = concat(results, axis=1, copy=False)
         result.columns = self.columns
+        # GH#44414: call finalize to retain metadata for the DataFrame or Series
```
Can remove the comment

skwirskj replied:
removed

phofl commented on pandas/tests/frame/methods/test_astype.py:
```diff
@@ -714,6 +714,15 @@ def test_astype_noncontiguous(self, index_slice):
+        df = DataFrame({"a": [0, 1, 2], "b": [3, 4, 5]})
+        df.attrs["Location"] = "Michigan"
+
+        df2 = df.astype({"a": float})
```
Please call result

phofl commented on pandas/tests/frame/methods/test_astype.py:
```diff
@@ -714,6 +714,15 @@ def test_astype_noncontiguous(self, index_slice):
         expected = df.iloc[index_slice]
         tm.assert_frame_equal(result, expected, check_dtype=False)
 
+    def test_astype_retain_attrs(self):
```
Could you check if we have tests for the other cases. If not, please parametrize here

skwirskj replied:
Just to clarify, we would want a case for changing all columns' data type correct?

phofl replied:
No, there are different options of parametrizing astype, a data type, data type represented as string or a dict of one of both

skwirskj replied:
Oh ok, I see, I will double-check for this and add any tests if necessary.

skwirskj replied:
Added the parametrization

phofl commented:
Couple of comments

skwirskj committed:
BUG: GH#44424 Updated whatsnew and tests

Changed description of bug fix to better explain what is being changed. Added more explicit varibale names in astype test to demonstrate the result and expected values.

skwirskj committed:
BUG: GH#44414 parametrized test_astype_retain_attrs in test_astype

jreback commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -636,7 +636,7 @@ Conversion
 - Bug in :func:`to_datetime` with ``arg:xr.DataArray`` and ``unit="ns"`` specified raises TypeError (:issue:`44053`)
 - Bug in :meth:`DataFrame.convert_dtypes` not returning the correct type when a subclass does not overload :meth:`_constructor_sliced` (:issue:`43201`)
--
+- Bug in :meth:`DataFrame.astype` not propagating ``attrs`` from the original :class:`DataFrame` (:issue:`44414`)
```
can you move this next to the other notes about attr changes

skwirskj replied:
Yes I can, I only found mentions of Attributes being changed within I/O and sparse section. Should I move it to one of those? Thank you in advance for the help

jreback reviewed (changes requested):
looks fine, small comment. we have a comprehensive test for methods, odd that this doesn't need changing (as it would not xpass)

skwirskj committed:
Merge remote-tracking branch 'upstream/master' into 44414-DataFrameAttr

jreback commented on pandas/tests/frame/methods/test_astype.py:
```diff
@@ -714,6 +714,19 @@ def test_astype_noncontiguous(self, index_slice):
         tm.assert_frame_equal(result, expected, check_dtype=False)
 
+    @pytest.mark.parametrize(
+        "dtype", ["float", "int32", "int64", "string", "int16", "float32"]
```
can you use one of the fixtures in pandas/conftest.py instead

jreback reviewed (changes requested):
add a test for series as well. merge master and ping on green.

skwirskj committed:
BUG: GH#44414 Added fixture use in test_astype.py

Added fixture use in both test_astype_retain_attrs functions in the DataFrame and Series testing files.

skwirskj committed:
Merge remote-tracking branch 'upstream/master' into 44414-DataFrameAttr

jreback commented:
thanks @skwirskj

skwirskj commented:
No problem @jreback thank you for your help!
